crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -578,9 +703,9 @@ mod tests {
         assert_eq!(count_loads(main.entry_block(), &main.dfg), 0);
         assert_eq!(count_loads(b1, &main.dfg), 0);
 
-        // All stores should be removed
-        assert_eq!(count_stores(main.entry_block(), &main.dfg), 0);
-        assert_eq!(count_stores(b1, &main.dfg), 0);
+        // Neither store is removed since they are each the last in the block and there are multiple blocks
```

Comment:
So even though we can resolve v3 and v4 is this due to v0 potentially being an alias?

Reply:
The mem2reg pass only tries to remove the last store in a block if it is also the only block in the function.

We could expand the analysis across multiple blocks and we'd have to check if there were no unoptimized loads between that store and every successor of that block. I think this would be too much bookkeeping since it requires us to track block successors now as well and it is unneeded since after flattening there will only be one block remaining anyway. For brillig functions which do not flatten (in the future when we apply optimizations to them), we still try to optimize out loads as normal but there may be some stores remaining since only multiple stores within the same block would be removed.